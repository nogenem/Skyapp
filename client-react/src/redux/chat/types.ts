import { TMessageType } from '~/constants/message_types';
import { TUserStatus } from '~/constants/user_status';

interface IOtherUser {
  _id: string;
  nickname: string;
  thoughts: string;
  status: TUserStatus;
  online: boolean;
  channel_id?: string;
}

interface IOtherUsers {
  [_id: string]: IOtherUser;
}

interface IMember {
  user_id: string;
  is_adm: boolean;
  last_seen: Date;
}

interface IChannel {
  _id: string;
  name: string;
  is_group: boolean;
  members: IMember[];
  other_member_idx?: number;
  unread_msgs: number;
  lastMessage?: IMessage;
}

interface IChannels {
  [_id: string]: IChannel;
}

interface IAttachment {
  originalName: string;
  size: number;
  path: string;
  mimeType: string;
  imageDimensions?: { width: number; height: number };
}

interface IMessage {
  _id: string;
  channel_id: string;
  from_id?: string;
  body: string | IAttachment;
  type: TMessageType;
  createdAt: Date;
  updatedAt: Date;
}

interface IInitialData {
  users: IOtherUsers;
  channels: IChannels;
}

interface IActiveChannelInfo {
  _id: string;
  messages: IMessage[];
  totalMessages: number;
  queue: IMessage[];
}

interface INewGroupCredentials {
  name: string;
  members: string[];
  admins: string[];
}

interface IUpdateGroupCredentials {
  channel_id: string;
  name: string;
  members: string[];
  admins: string[];
}

interface ILeaveGroupCredentials {
  channel_id: string;
}

interface IFetchMessagesCredentials {
  channel_id: string;
  offset: number;
  limit?: number;
  sort?: string;
}

interface ISendMessageCredentials {
  channel_id: string;
  body: string;
}

type TChatState = {
  users: IOtherUsers;
  channels: IChannels;
  activeChannelInfo?: IActiveChannelInfo;
};

enum EChatActions {
  SET_INITIAL_DATA = '@chat/SET_INITIAL_DATA',
  SET_USER_ONLINE = '@chat/SET_USER_ONLINE',
  ADD_OR_UPDATE_CHANNEL = '@chat/ADD_OR_UPDATE_CHANNEL',
  SET_ACTIVE_CHANNEL = '@chat/SET_ACTIVE_CHANNEL',
  REMOVE_CHANNEL = '@chat/REMOVE_CHANNEL',
  ADD_NEW_USER = '@chat/ADD_NEW_USER',
  ADD_MESSAGES = '@chat/ADD_MESSAGES',
  SET_LATEST_MESSAGE = '@chat/SET_LATEST_MESSAGE',
  ADD_TO_MESSAGES_QUEUE = '@chat/ADD_TO_MESSAGES_QUEUE',
  REMOVE_FROM_MESSAGES_QUEUE = '@chat/REMOVE_FROM_MESSAGES_QUEUE',
  SET_LAST_SEEN = '@chat/SET_LAST_SEEN',
  SET_USER_STATUS = '@chat/SET_USER_STATUS',
  SET_USER_THOUGHTS = '@chat/SET_USER_THOUGHTS',
}

interface IChatActionType<T, P> {
  type: T;
  payload: P;
}

type TChatAction =
  | IChatActionType<typeof EChatActions.SET_INITIAL_DATA, IInitialData>
  | IChatActionType<
      typeof EChatActions.SET_USER_ONLINE,
      { _id: string; value: boolean }
    >
  | IChatActionType<typeof EChatActions.ADD_OR_UPDATE_CHANNEL, IChannel>
  | IChatActionType<
      typeof EChatActions.SET_ACTIVE_CHANNEL,
      { _id: string | undefined }
    >
  | IChatActionType<typeof EChatActions.REMOVE_CHANNEL, { channelId: string }>
  | IChatActionType<typeof EChatActions.ADD_NEW_USER, IOtherUser>
  | IChatActionType<
      typeof EChatActions.ADD_MESSAGES,
      { messages: IMessage[]; totalMessages: number; atTop: boolean }
    >
  | IChatActionType<typeof EChatActions.SET_LATEST_MESSAGE, IMessage>
  | IChatActionType<typeof EChatActions.ADD_TO_MESSAGES_QUEUE, IMessage>
  | IChatActionType<typeof EChatActions.REMOVE_FROM_MESSAGES_QUEUE, IMessage>
  | IChatActionType<
      typeof EChatActions.SET_LAST_SEEN,
      {
        channel_id: string;
        user_id: string;
        last_seen: string;
      }
    >
  | IChatActionType<
      typeof EChatActions.SET_USER_STATUS,
      {
        user_id: string;
        newStatus: TUserStatus;
      }
    >
  | IChatActionType<
      typeof EChatActions.SET_USER_THOUGHTS,
      {
        user_id: string;
        newThoughts: string;
      }
    >;

export type {
  TChatState,
  TChatAction,
  IInitialData,
  IOtherUsers,
  IOtherUser,
  IChannel,
  IChannels,
  IMember,
  IMessage,
  IAttachment,
  INewGroupCredentials,
  IUpdateGroupCredentials,
  ILeaveGroupCredentials,
  IFetchMessagesCredentials,
  ISendMessageCredentials,
};
export { EChatActions };
