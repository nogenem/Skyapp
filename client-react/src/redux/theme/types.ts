type TThemeMode = 'dark' | 'light';

type TThemeState = {
  mode?: TThemeMode;
};

enum EThemeActions {
  MODE_SWITCHED = '@theme/MODE_SWITCHED',
}

interface IThemeActionType<T, P> {
  type: T;
  payload: P;
}

type TThemeAction = IThemeActionType<
  typeof EThemeActions.MODE_SWITCHED,
  TThemeMode
>;

export type { TThemeMode, TThemeState, TThemeAction };
export { EThemeActions };
