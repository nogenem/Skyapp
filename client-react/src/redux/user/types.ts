interface User {
  _id: string;
  token?: string;
}

type UserState = User;

enum EUserActions {
  SIGNED_IN = '@user/SIGNED_IN',
  SIGNED_OUT = '@user/SIGNED_OUT',
}

interface UserActionType<T, P> {
  type: T;
  payload: P;
}

type UserAction =
  | UserActionType<typeof EUserActions.SIGNED_IN, User>
  | UserActionType<typeof EUserActions.SIGNED_OUT, null>;

export type { User, UserState, UserAction };
export { EUserActions };
