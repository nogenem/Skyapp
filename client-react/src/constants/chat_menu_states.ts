export const MENU_STATES = {
  MAIN: 1,
  CHANGING_LANGUAGE: 2,
  CHANGING_USER_STATUS: 3,
  CHANGING_USER_THOUGHTS: 4,
} as const;
type TMenuStates = typeof MENU_STATES[keyof typeof MENU_STATES];

export type { TMenuStates };
