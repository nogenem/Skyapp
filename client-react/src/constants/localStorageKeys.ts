export const LOCAL_STORAGE_TOKEN = 'skyapp-jwt';
export const LOCAL_STORAGE_THEME_MODE = 'skyapp-theme-mode';
