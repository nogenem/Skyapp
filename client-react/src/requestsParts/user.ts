import { TUserStatus } from '~/constants/user_status';

interface IChangeStatusRequestBody {
  newStatus: TUserStatus;
}

interface IChangeThoughtsRequestBody {
  newThoughts: string;
}

export type { IChangeStatusRequestBody, IChangeThoughtsRequestBody };
