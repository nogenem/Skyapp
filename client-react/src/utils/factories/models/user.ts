import merge from 'deepmerge';
import faker from 'faker/locale/en_US';

import { USER_STATUS } from '~/constants/user_status';
import type { IUser } from '~/redux/user/types';

export default (override?: Partial<IUser>): IUser =>
  merge(
    {
      _id: faker.datatype.uuid(),
      nickname: faker.internet.userName(),
      email: faker.internet.email(),
      confirmed: Math.random() < 0.5,
      status: faker.helpers.randomize(Object.values(USER_STATUS)),
      thoughts: faker.lorem.sentence(),
      token: faker.datatype.uuid(),
    } as IUser,
    override || {},
  );
