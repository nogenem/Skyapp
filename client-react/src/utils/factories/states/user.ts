import { initialState } from '~/redux/user/reducer';
import type { TUserState } from '~/redux/user/types';

import { user as userFactory } from '../models';

interface IOptions {
  useConstValues: boolean;
  useInitialState: boolean;
}

export default (
  override?: Partial<TUserState>,
  options?: Partial<IOptions>,
): TUserState =>
  options?.useInitialState
    ? initialState
    : userFactory(override, { useConstValues: options?.useConstValues });

export type { IOptions };
