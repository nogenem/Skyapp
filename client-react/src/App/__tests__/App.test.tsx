import React from 'react';
import { Provider } from 'react-redux';

import { render, waitForElementToBeRemoved } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { LOCAL_STORAGE_TOKEN } from '~/constants/localStorageKeys';
import type { IAppState } from '~/redux/store';
import { ITokenCredentials } from '~/redux/user/types';

import { App, UnconnectedApp } from '../index';

jest.mock('../../redux/user/actions', () => ({
  validateToken: () => () => Promise.resolve(),
}));

const VALID_TOKEN = '123456789';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const emptyState: Partial<IAppState> = {};
const renderWithRedux = (
  ui: React.ReactNode,
  initialState: Partial<IAppState> = emptyState,
) => {
  const store = mockStore(initialState);
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
  };
};

describe('Connected App', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders correctly', async () => {
    const { container, queryByTestId } = renderWithRedux(<App />);

    await waitForElementToBeRemoved(() => queryByTestId(/spinner_div/i), {
      container: container as HTMLElement,
    }); // wait for the Spinner to disappear

    expect(container).toMatchSnapshot();
  });
});

describe('Unconnected App', () => {
  it('Trys to confirm token', async () => {
    const credentials: ITokenCredentials = { token: VALID_TOKEN };
    const validateToken = jest.fn(() => Promise.resolve());

    localStorage.setItem(LOCAL_STORAGE_TOKEN, VALID_TOKEN);
    const { container, queryByTestId } = renderWithRedux(
      <UnconnectedApp validateToken={validateToken} />,
    );

    await waitForElementToBeRemoved(() => queryByTestId(/spinner_div/i), {
      container: container as HTMLElement,
    }); // wait for the Spinner to disappear

    expect(validateToken).toHaveBeenCalledTimes(1);
    expect(validateToken).toHaveBeenCalledWith(credentials);
  });
});
