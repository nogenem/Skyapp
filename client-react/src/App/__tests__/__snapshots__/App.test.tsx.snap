// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected App renders correctly 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary mui-fixed MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <h6
        class="MuiTypography-root makeStyles-title-3 MuiTypography-h6 MuiTypography-noWrap"
      >
        SkyApp
      </h6>
      <button
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
        data-testid="open_lang_menu"
        tabindex="0"
        title="Messages:Choose a language"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.87 15.07l-2.54-2.51.03-.03c1.74-1.94 2.98-4.17 3.71-6.53H17V4h-7V2H8v2H1v1.99h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11.76-2.04zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2l-4.5-12zm-2.62 7l1.62-4.33L19.12 17h-3.24z"
            />
          </svg>
          <span
            class="MuiTypography-root makeStyles-language-4 MuiTypography-button"
          >
            Languages:en-US
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Messages:Toggle dark/light theme"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        data-testid="toggle_theme_btn"
        tabindex="0"
        title="Messages:Toggle dark/light theme"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
  <main
    class="makeStyles-content-1"
  />
</div>
`;
